bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/ICalendarHandlerFactory.java

```diff
@@ -67,6 +84,61 @@ public boolean supportsThingType(ThingTypeUID thingTypeUID) {
         if (!supportsThingType(thingTypeUID)) {
             return null;
         }
-        return new ICalendarHandler(thing, sharedHttpClient, eventPublisher);
+        if (thingTypeUID.equals(THING_TYPE_CALENDAR)) {
+            if (thing instanceof Bridge) {
+                return new ICalendarHandler((Bridge) thing, sharedHttpClient, eventPublisher);
+            } else {
+                // Migration needs to be done asynchronously. Using thread pool for common things.
+                ThingUID uidToCopy = thing.getUID();
+                ExecutorService threadPool = ThreadPoolManager.getPool(ThreadPoolManager.THREAD_POOL_NAME_COMMON);
+                threadPool.execute(new ReregisterThingRunnable(uidToCopy));
+                return null;
+            }
+        } else if (thingTypeUID.equals(THING_TYPE_FILTERED_EVENTS)) {
+            return new EventFilterHandler(thing);
+        }
+        return null;
+    }
+
+    /**
+     * This Runnable "upgrades" the thing "calendar" to the bridge of same type. To make sure the thing isn't in use, a
+     * sleep blocks before doing actual work.
+     *
+     * @author Michael Wodniok - Initial contribution
+     */
+    private class ReregisterThingRunnable implements Runnable {
+        private final ThingUID thingUID;
+
+        public ReregisterThingRunnable(ThingUID uid) {
+            thingUID = uid;
+        }
+
+        @Override
+        public void run() {
+            logger.info("Converting Thing {} to Bridge. This will remove the item and readd it. Please stand by.",
```

Comment:
Logging to info should be used rarely. This could be debug. Same for below.

Reply:
This is intentionally info as it changes the users configuration. This should happen exactly once, because the handler only converted once. In that time the item is shortly unusable. Should we really hide the change to users configuration? It could lead to confusion because something changes without any notification - this way a user may find out, why something in his configuration changed.